### PR TITLE
Add Symfony-based rsync file copier alternative and end-to-end functional test coverage

### PR DESCRIPTION
**Pull request by TravisCarden**
This adds a Symfony Filesystem-based file copier that's automatically used on hosts (like Windows) that don't have `rsync` available. It also adds full, end-to-end functional testing of the entire project. Functional tests are currently restricted to the happy path. We can decide later which edge error/edge conditions are important enough to test.

### CODE REVIEW

File: src/Infrastructure/Process/FileCopier/FileCopierFactory.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Infrastructure\Process\FileCopier;
+
+use Symfony\Component\Process\ExecutableFinder;
+
+/**
+ * @internal
```
Comment (TravisCarden) [inline src/Infrastructure/Process/FileCopier/FileCopierFactory.php:8]: Full docblocks are on the interfaces.

File: tests/Functional/EndToEndTest.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        self::createTestEnvironment();
+        self::initializeComposerJson();
+    }
+
+    public static function tearDownAfterClass(): void
+    {
+        self::removeTestEnvironment();
+    }
+
+    protected static function runFrontScript(array $args, string $cwd = __DIR__): Process
```
Comment (TravisCarden) [inline tests/Functional/EndToEndTest.php:23]: Practically all the supporting methods in tests, including assertions, have to be static for consistency or because they have to be callable by static methods like `::tearDownAfterClass()` that inherit directly from PHPUnit.

File: src/Infrastructure/Process/FileCopier/SymfonyFileCopier.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * @throws \PhpTuf\ComposerStager\Exception\DirectoryNotFoundException
+     */
+    private function createIterator(string $from): RecursiveCallbackFilterIterator
+    {
+        try {
+            $directoryIterator = new RecursiveDirectoryIterator($from);
+        } catch (UnexpectedValueException $e) {
+            throw new DirectoryNotFoundException($from, 'The "copy from" directory does not exist at "%s"');
```
Comment (phenaproxima) [inline src/Infrastructure/Process/FileCopier/SymfonyFileCopier.php:48]: Passing the previous exception as well can aid with debugging.

File: tests/Functional/EndToEndTest.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Tests\Functional;
+
+use Symfony\Component\Process\Process;
+
+/**
+ * @coversNothing
```
Comment (phenaproxima) [inline tests/Functional/EndToEndTest.php:8]: Kinda makes me wish there was a `@coversEverything` tag, since that would probably be more accurate for an end-to-end test. =P

File: tests/Functional/TestCase.php [hunk 5]
```diff
@@ -2,22 +2,101 @@
...
+    protected static function assertActiveAndStagingDirectoriesNotSame(): void
+    {
+        self::assertNotSame(
+            '',
+            self::activeAndStagingDirectoriesDiff(),
+            'Active and staging directories are not the same.'
+        );
+    }
+
+    protected static function activeAndStagingDirectoriesDiff(): string
```
Comment (phenaproxima) [inline tests/Functional/TestCase.php:92]: This method name should maybe begin with a verb, to make it clear what it does.

**Review by phenaproxima — APPROVED** [review]

File: src/Infrastructure/Process/FileCopier/SymfonyFileCopier.php (hunk 3 above)
Comment (TravisCarden) [reply src/Infrastructure/Process/FileCopier/SymfonyFileCopier.php:48]: Thanks, I keep missing forgetting to do that. I'll see if I can write a static analysis check for it for future.
